Tasks/SonarQubePostTest/CodeAnalysisFilePathComputation.ps1

```diff
@@ -0,0 +1,170 @@
+    Write-Verbose -Verbose "GetRelativeFilePath: Returning finalFilePath:$finalFilePath"
+
+    #save data into cache so next time we don't have to compute
+    $ComponentKeyAndRelativePathCache.Add($component, $finalFilePath)
+
+    return $finalFilePath
+}
+
+function ProcessSonarCodeAnalysisReport
+{
+    param([string][ValidateNotNullOrEmpty()]$agentBuildDirectory)
+
+    $sonarReportFolderPath = [System.IO.Path]::Combine($agentBuildDirectory, ".sonarqube", "out", ".sonar")
+    $sonarReportFilePath = [System.IO.Path]::Combine($sonarReportFolderPath, "sonar-report.json")
+
+    $sonarReportFilePathProcessed = [System.IO.Path]::Combine($sonarReportFolderPath, "code-analysis-report.json")
+    
+    if (![System.IO.File]::Exists($sonarReportFilePath))
+    {
+        Write-Host "ProcessSonarCodeAnalysisReport: $sonarReportFilePath does not exist! Returning.."
+        return;
+    }
+
+    #read sonar-report.json file as a json object
+    $json = Get-Content -Raw $sonarReportFilePath | ConvertFrom-Json
+    Write-Verbose -Verbose "ProcessSonarCodeAnalysisReport: Total issues: $($json.issues.Count)"
+
+    ConstructComponentKeyAndPathMap $json
+
+    # '@' makes sure the result set is returned as an array
+    $newIssues = @($json.issues | Where {$_.isNew -eq $true})
+    Write-Verbose -Verbose "ProcessSonarCodeAnalysisReport: Total new issues: $($newIssues.Count)"
+
+    if ($($newIssues.Count) -gt 0)
+    {
+        foreach ($issue in $newIssues)
+        {
+            $filePath = GetRelativeFilePath $($issue.component)
+
+            #add a new property in json which stores the file path so it can be consumed directly
+            Add-Member -InputObject $issue -MemberType NoteProperty -Name relativePath -Value $filePath
+        }
+
+        #save the results into output file
+        $newIssues | ConvertTo-Json | Set-Content -Path $sonarReportFilePathProcessed
+    }
+}
+
+#creates a mapping of msbuild project guid and the path on disk using ProjectInfo.xml file
+function CreateProjectGuidAndPathMap
+{
+    param([string][ValidateNotNullOrEmpty()]$agentBuildDirectory)
+
+    $parentFolder = [System.IO.Path]::Combine($agentBuildDirectory, ".sonarqube", "out")
+    $parentFolderItem = Get-Item $parentFolder
+    $directories = $parentFolderItem.GetDirectories()     
+
+    foreach ($directory in $directories)
+    {
+        $projectInfoFilePath = [System.IO.Path]::Combine($directory.FullName, "ProjectInfo.xml")
+        
+        #Write-Host $projectInfoFilePath
+        
+        if ([System.IO.File]::Exists($projectInfoFilePath))
+        {
+            Write-host "CreateProjectGuidAndPathMap: Processing project info file: $projectInfoFilePath"
+            [xml]$xmlContent = Get-Content $projectInfoFilePath
+
+            if ($xmlContent -ne $null)
+            {
+                $ProjectGuidAndFilePathMap.Add($xmlContent.ProjectInfo.ProjectGuid, $xmlContent.ProjectInfo.FullPath)
```

Comment:
It's possible that the same project guid will appear in more than once MSBuild project file, or that the project guid is missing. The MSBuild Runner handles these cases by excluding projects with duplicates/missing ids from analysis.
Consequently, there shouldn't be any issues reported in the sonar-runner.json file for such projects.
However, you're parsing the ProjectInfo.xml files on disc directly so you need to check that this code doesn't fail if it encounters multiple/missing ids.